linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,303 @@
+        {
+            LogError("Failure allocating THREADPOOL structure");
+        }
+        else
+        {
+            result->max_thread_count = MAX_THREAD_COUNT;
+
+            // Create a list of threads
+            result->thread_handle_list = malloc(sizeof(THREAD_HANDLE)*result->max_thread_count);
+            if (result->thread_handle_list == NULL)
+            {
+                LogError("Failure allocating THREAD array structure %zu", sizeof(THREAD_HANDLE)*result->max_thread_count);
+            }
+            else
+            {
+                result->list_index = 0;
+                result->srw_lock = srw_lock_create(false, "");
+                if (result->srw_lock == NULL)
+                {
+                    LogError("Failure srw_lock_create");
+                }
+                else
+                {
+                    result->thread_count = 0;
+                    result->task_count = 0;
+                    result->quitting = 0;
+                    (void)interlocked_exchange(&result->pool_state, POOL_STATE_UNINIT);
+
+                    result->task_list = NULL;
+                    result->task_list_last = NULL;
+                    goto all_ok;
+                }
+                free(result->thread_handle_list);
+            }
+            free(result);
+            result = NULL;
+        }
+    }
+all_ok:
+    return result;
+}
+
+void threadpool_destroy(THREADPOOL_HANDLE threadpool)
+{
+    if (threadpool == NULL)
+    {
+        // do nothing
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p", threadpool);
+    }
+    else
+    {
+        interlocked_exchange(&threadpool->quitting, 1);
+
+        for (size_t index = 0; index < threadpool->list_index; index++)
+        {
+            int dont_care;
+            ThreadAPI_Join(threadpool->thread_handle_list[index], &dont_care);
+        }
+
+        free(threadpool->thread_handle_list);
+
+        srw_lock_destroy(threadpool->srw_lock);
+        free(threadpool);
+    }
+}
+
+int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCTION work_function, void* work_function_context)
+{
+    int result;
+    if (threadpool == NULL || work_function == NULL)
+    {
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p, THREADPOOL_TASK_FUNC work_function: %p", threadpool, work_function);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        THREADPOOL_TASK* task_item = malloc(sizeof(THREADPOOL_TASK));
+        if (task_item == NULL)
+        {
+            LogError("Failure allocating THREADPOOL_TASK structure");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            task_item->next = NULL;
+            task_item->task_param = work_function_context;
+            task_item->task_func = work_function;
+
+            // Need to add a lock here
+            srw_lock_acquire_exclusive(threadpool->srw_lock);
+
+            if (threadpool->task_list == NULL)
```

Comment:
>task_list [](http://example.com/codeflow?start=28&length=9)

This is clearly accessed from different threads so I think we need interlocked operations on it.

Reply:
This is inside the srw_lock, so it should be protected correct?